applications/vessel-history/src/features/map/map.hooks.ts

```diff
@@ -0,0 +1,165 @@
+//       }
+//       setHoverCenter(center)
+//       if (rulersEditing === true) {
+//         dispatch(moveCurrentRuler(center))
+//       }
+//     },
+//     [dispatch, rulersEditing]
+//   )
+//   return { onMapMove, hoverCenter }
+// }
+// This is a convenience hook that returns at the same time the portions of the store we interested in
+// as well as the functions we need to update the same portions
+export const useGeneratorsConnect = () => {
+  const dispatch = useDispatch()
+  return {
+    globalConfig: useSelector(selectGlobalGeneratorsConfig),
+    generatorsConfig: useSelector(selectGeneratorsConfig),
+    updateGenerator: (payload: UpdateGeneratorPayload) => dispatch(updateGenerator(payload)),
+  }
+}
+type SetMapCoordinatesArgs = Pick<ViewportProps, 'latitude' | 'longitude' | 'zoom'>
+type UseViewport = {
+  viewport: MapCoordinates
+  onViewportChange: (viewport: ViewportProps) => void
+  setMapCoordinates: (viewport: SetMapCoordinatesArgs) => void
+}
+
+export type LatLon = {
+  latitude: number
+  longitude: number
+}
+export interface Viewport extends LatLon {
+  latitude: number
+  longitude: number
+  zoom: number
+}
+// export const useViewportConnect = () => {
+//   const dispatch = useDispatch()
+//   const { zoom, latitude, longitude } = useSelector(selectViewport)
+//   const dispatchViewport = (newViewport: Partial<Viewport>) =>
+//     dispatch(updateQueryParams(newViewport))
+//   return { zoom, latitude, longitude, dispatchViewport }
+// }
+
+// export function useDebouncedViewport(
+//   urlViewport: MapCoordinates,
+//   callback: (viewport: MapCoordinates) => void
+// ): UseViewport {
+//   const [viewport, setViewport] = useState<MapCoordinates>(urlViewport)
+//   const debouncedViewport = useDebounce<MapCoordinates>(viewport, 400)
+
+//   const setMapCoordinates = useCallback((viewport: SetMapCoordinatesArgs) => {
+//     setViewport({ ...viewport })
+//   }, [])
+
+//   const onViewportChange = useCallback((viewport: ViewportProps) => {
+//     const { latitude, longitude, zoom } = viewport
+//     setViewport({ latitude, longitude, zoom })
+//   }, [])
+
+//   // Updates local state when url changes
+//   useEffect(() => {
+//     const { latitude, longitude, zoom } = viewport
+//     if (
+//       urlViewport &&
+//       (urlViewport?.latitude !== latitude ||
+//         urlViewport?.longitude !== longitude ||
+//         urlViewport?.zoom !== zoom)
+//     ) {
+//       setViewport({ ...urlViewport })
+//     }
+//     // eslint-disable-next-line react-hooks/exhaustive-deps
+//   }, [urlViewport])
+
+//   // Sync the url with the local state debounced
+//   useEffect(() => {
+//     if (debouncedViewport && callback) {
+//       callback(debouncedViewport)
+//     }
+//     // eslint-disable-next-line react-hooks/exhaustive-deps
+//   }, [debouncedViewport])
+
+//   return { viewport, onViewportChange, setMapCoordinates }
+// }
+
+// export function useViewport(): UseViewport {
```

Comment:
This always was an 🤯 let's talk about the latest working solution I found using [an atom](https://github.com/GlobalFishingWatch/frontend/blob/master/applications/fishing-map/src/features/map/map-viewport.hooks.ts#L26)

Reply:
I'll try that